### PR TITLE
Refactor webhooks to reduce code duplication

### PR DESCRIPTION
**Pull request by cornelk**
This WIP pull request shows a WIP refactoring of `get*IssuesPayload` functions to reduce the code duplication and also adds unit tests. Feedback is welcome.

### CODE REVIEW

File: modules/webhook/general.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
+package webhook
```
Comment (lunny) [inline modules/webhook/general.go:5]: copyright head missed

**Comment by cornelk** [comment]
Ready for review

File: modules/webhook/general.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+	return text, color
+}
+
+func getIssueCommentPayloadInfo(p *api.IssueCommentPayload, linkFormatter linkFormatter) (string, string, int) {
+	senderLink := linkFormatter(setting.AppURL+p.Sender.UserName, p.Sender.UserName)
+	repoLink := linkFormatter(p.Repository.HTMLURL, p.Repository.FullName)
+	issueTitle := fmt.Sprintf("#%d %s", p.Issue.Index, p.Issue.Title)
+
+	var text, typ, titleLink string
```
Comment (davidsvantesson) [inline modules/webhook/general.go:158]: typ->type

**Review by davidsvantesson — CHANGES_REQUESTED** [review]

File: modules/webhook/general.go (hunk 2 above)
Comment (lunny) [reply modules/webhook/general.go:158]: It's a keyword of golang, you cannot use that.

**Review by lunny — APPROVED** [review]

File: modules/webhook/general_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package webhook
+
+import api "code.gitea.io/gitea/modules/structs"
```
Comment (6543) [inline modules/webhook/general_test.go]:
```suggestion
import (
	api "code.gitea.io/gitea/modules/structs"
)
```

**Review by 6543 — APPROVED** [review]

File: modules/webhook/general_test.go (hunk 3 above)
Comment (6543) [reply modules/webhook/general_test.go]: only false code format on test found

**Review by 6543 — COMMENTED** [review]
this will change the messages a bit!!

**Review by 6543 — CHANGES_REQUESTED** [review]
on getDingtalkIssuesPayload `p.Index` is not there anymore

EDIT: `issueTitle := fmt.Sprintf("#%d %s", p.Index, p.Issue.Title)` wich is in `titleLink`

so only https://github.com/go-gitea/gitea/pull/9422#discussion_r361733272 would be nice to have

File: modules/webhook/general.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	senderLink := linkFormatter(setting.AppURL+p.Sender.UserName, p.Sender.UserName)
+	repoLink := linkFormatter(p.Repository.HTMLURL, p.Repository.FullName)
+	issueTitle := fmt.Sprintf("#%d %s", p.Index, p.Issue.Title)
+	titleLink := linkFormatter(fmt.Sprintf("%s/issues/%d", p.Repository.HTMLURL, p.Index), issueTitle)
+	var text string
+	color := yellowColor
+
+	switch p.Action {
+	case api.HookIssueOpened:
+		text = fmt.Sprintf("[%s] Issue opened by %s", repoLink, senderLink)
```
Comment (zeripath) [inline modules/webhook/general.go]: We really should have the index of the issue listed here
```suggestion
		text = fmt.Sprintf("[%s] Issue #%d opened by %s", repoLink, p.index, senderLink)
```

**Comment by 6543** [comment]
@cornelk for `getIssuesPayloadInfo` the case `HookIssueOpened` should also have issue index

File: modules/webhook/general.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	case api.HookIssueClosed:
+		text = fmt.Sprintf("[%s] Issue closed: %s by %s", repoLink, titleLink, senderLink)
+		color = redColor
+	case api.HookIssueReOpened:
+		text = fmt.Sprintf("[%s] Issue re-opened: %s by %s", repoLink, titleLink, senderLink)
+	case api.HookIssueEdited:
+		text = fmt.Sprintf("[%s] Issue edited: %s by %s", repoLink, titleLink, senderLink)
+	case api.HookIssueAssigned:
+		text = fmt.Sprintf("[%s] Issue assigned to %s: %s by %s", repoLink,
+			linkFormatter(setting.AppURL+p.Issue.Assignee.UserName, p.Issue.Assignee.UserName),
```
Comment (6543) [inline modules/webhook/general.go]: are you sure `setting.AppURL` end with a  `/` ? can you use path.Join()

File: modules/webhook/general.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	return text
+}
+
+// htmlLinkFormatter creates a HTML link
+func htmlLinkFormatter(url string, text string) string {
+	return fmt.Sprintf(`<a href="%s">%s</a>`, url, html.EscapeString(text))
+}
+
+func getIssuesPayloadInfo(p *api.IssuePayload, linkFormatter linkFormatter) (string, string, string, int) {
+	senderLink := linkFormatter(setting.AppURL+p.Sender.UserName, p.Sender.UserName)
```
Comment (6543) [inline modules/webhook/general.go]: same here https://github.com/go-gitea/gitea/pull/9422/files#r361738973

**Review by zeripath — DISMISSED** [review]
As per @6543 review comments

**Comment by cornelk** [comment]
@zeripath @6543 fix pushed

**Comment by 6543** [comment]
@cornelk can you ''update branch"?

**Review by 6543 — APPROVED** [review]

**Pull request merged**